\documentclass[10pt]{article}
\usepackage[T1]{fontenc}
\usepackage{tabularx}
\usepackage{parskip}
\usepackage{float}
\usepackage{hyperref}
\usepackage{amsmath}
\usepackage{xcolor}
\usepackage{graphicx}
\usepackage{soul}
\usepackage[margin=1in]{geometry}

\hypersetup{
    colorlinks,
    linkcolor={red!50!black},
    citecolor={blue!50!black},
    urlcolor={blue!80!black}
}

\setlength{\parindent}{0cm}

\begin{document}

\title{\textbf{KINC v0.1 Specification}\\
\vspace{1mm}
\includegraphics[width=8cm,height=4cm]{KINClogo.png}
\\Knowledge Independent Network Construction}
\author{Joshua Burns\thanks{Dept. of Horticulture, Washington State University}, Stephen Ficklin\footnotemark[1]}
\maketitle

\newpage
\tableofcontents

\newpage
\section{Introduction}

KINC is designed for use in construction of biological networks, specifically, 
gene co-expression networks. KINC performs three major steps:  1) construction
of a similarity matrix of pair-wise expression correlations, 2) thresholding of 
the similarity matrix to form an adjaceny matrix, 3) export of the adjaceny 
matrix to form a tab-delimited network file.  

This document provides an overview for the data structures and file formats used 
by KINC. 

\hl{Can we describe here how the command-line for KINC will work and how
use of the classes described below translates into command-line arguments?}

\newpage
\section{Abstract Classes}

The following abstract classes serve as the base for all structures within the 
kinc program.  All other classes should inherit from these. This design will 
allow for dynamic addition of user implemented plugins in the future, because
the functions of the abstract classes are clearly defined.

\subsection{KINCData}

Th KINCData class is responsible for reading and writing of KINC data files.  
Classes that inherit from the KINCData class are responsible for implementing 
the generic functions exposed by this class.  This includes reading, writing, 
merging, exporing, indexing and querying. Typically, child classes provide 
importers that read commonly used file formats into their own binary file 
format and exporters to convert back to those same file formats.

\subsection{KINCAnalytic}

The KINCAnalytic class is responsible for taking in one or more KINCData objects
and employ an algorithm such as a statistical test to produce one or more 
new KINCData objects.   A KINCAnalytic 

\newpage
\section{KINCData Classes}

\subsection{Expression}

The Expression class is responsible for manging gene expression-level data.  

\subsubsection{Properties}

\hl{Do we need any properties?}

\subsubsection{Constructor}

{\bfseries KINCData(int argc, char *argv[])}

\hl{We need to design how the functions of the class will receive arguments. 
 will we have a constructor that receives, parses and responds to errors
 for all functions?  Or should each function be responsible for checking 
 it's own arguments. I know we can't do that in the abstract class, but 
 we need to accomdate the behavior we settle on in our design so plugins
 are consistent.}

\subsubsection{Virtual Functions}

The following functions should be implemented by any plugin that creates
classes that inherits the KINCData class.

{\bfseries virtual void import() = 0}

This function reads a tab-delimited file.  Each line of this file
represents the gene expression levels of a single gene, transcript or probeset.
Each tab-separated value in a single line indicates the gene expression level 
for each sample. The expression level of a samples must be in the same 
order for every line.  The first line of the file may contain a tab-delimited
list of sample names, and a file may contain as many samples and genes as
desired.

{\bfseries virtual void export() = 0}

{\bfseries virtual void query() = 0}

{\bfseries virtual void merge() = 0}


\subsubsection{File Structure}

\begin{list}{}{}
\item[1.] Special Identifier that identifies this file as an expression file.
\item[2.] User defined name for expression data.
\item[3.] History of this data, where it came from.
\item[4.] Number of genes, then number of samples per gene.
\item[5.] List of all genes.
\item[6.] 2 Dimensional list of all gene samples, per gene.
\end{list}

\subsection{Correlation}

This is responsible for storing correlation data between genes.

\subsubsection{Correlation Binary File Format}
The following describes the format of the KINC correlation file. All 
multi-byte numbers are little-endian, regardless of the machine endianness.

\begin{tabular}{| l | l | l | l |}
  \hline
  {\bfseries Field } & {\bfseries Description } & {\bfseries Type } & {\bfseries Value } \\ 
  \hline			
  magic & The magic number that identifies this file as a correlation file. & char[5] & kcor\textbackslash1 \\ 
  \hline
  historyLen & The length of the header. & int32\_t & \\
  \hline
  nameLen & The length of the dataset name. & int32\_t & \\
  \hline
  name & The user defined name for this correlation dataset. & char[nameLen] & \\
  \hline
  historyLen & The length of history information. & int32\_t & \\
  \hline
  history & A string that describes the provenance of this file. & char[historyLen] & \\
  \hline
\end{tabular}
\hl{I like this type of table for describing the file format. I borrowed it 
from the BAM file specification}
\begin{list}{}{}
\item[1.] Special Identifier that identifies this file as a correlation file.
\item[2.] User defined name for correlation data.
\item[3.] History of this data, where it came from.
\item[4.] Number of genes, number of correlations per gene pair.
\item[5.] List of correlation methods used.
\item[5.] List of all genes.
\item[6.] Special lists of all gene correlations sorted by their correlation 
value. Each entry will have a correlation value, then a number representing how 
many gene correlations are this value, then the list of gene correlations. For 
each type of correlation used a list is made.
\item[7.] Special 2 Dimensional list of all gene correlations, special being a 
diagonal matrix instead of square.
\end{list}

\subsection{Network}

This is responsible for storing network data between genes.

\subsubsection{File Structure}

\begin{list}{}{}
\item[1.] Special Identifier that identifies this file as a network file.
\item[2.] User defined name for network data.
\item[3.] History of this data, where it came from.
\item[4.] Number of genes.
\item[5.] List of all genes.
\item[6.] Special 2 Dimensional list of all gene edges, special being a 
diagonal matrix instead of square.
\end{list}

\subsection{Annotation}

This is responsible for storing additional information for genes.

\subsubsection{File Structure}

\begin{list}{}{}
\item[1.] Special Identifier that identifies this file as an annotation file.
\item[2.] User defined name for annotation data.
\item[4.] Number of genes, number of annotations.
\item[5.] List of all genes.
\item[6.] List of all annotation types.
\item[7.] 2 Dimensional list of all gene annotations, per gene.
\end{list}

\newpage
\section{KINCAnalytic Classes}

\subsection{Pearson}

This takes an Expression BioData object and produces a Correlation BioData 
object. It uses the Pearson correlation statistical method for giving 
correlation values.

\subsection{Spearman}

This takes an Expression BioData object and produces a Correlation BioData 
object. It uses the Spearman correlation statistical method for giving 
correlation values.

\end{document}
