\documentclass[10pt]{article}
\usepackage[T1]{fontenc}
\usepackage{tabularx}
\usepackage{parskip}
\usepackage{float}
\usepackage{hyperref}
\usepackage{amsmath}
\usepackage{xcolor}
\usepackage{graphicx}
\usepackage{soul}
\usepackage{mdframed}
\usepackage{listings}
\usepackage[margin=1in]{geometry}

\hypersetup{
    colorlinks,
    linkcolor={red!50!black},
    citecolor={blue!50!black},
    urlcolor={blue!80!black}
}

\definecolor{btablecol}{rgb}{.9,.9,.9}
\mdfdefinestyle{btable}{
hidealllines=true,
backgroundcolor=btablecol,
innertopmargin=3mm,
innerbottommargin=3mm}

\definecolor{funclistbcol}{rgb}{.9,.9,1}
\definecolor{funclistlcol}{rgb}{.6,.6,1}
\mdfdefinestyle{functions}{
backgroundcolor=funclistbcol,
topline=false,
bottomline=false,
linecolor=funclistlcol,
linewidth=2mm}

\definecolor{lyellow}{rgb}{1,1,0.2}
\definecolor{lblue}{rgb}{0.8,0.8,1}
\providecommand{\stephen}[1]{\sethlcolor{lyellow}\hl{#1}}
\providecommand{\josh}[1]{\sethlcolor{lblue}\hl{#1}}

\definecolor{darkblue}{rgb}{0,0,0.6}
\providecommand{\h}[1]{\textcolor{darkblue}{#1}}

\setlength{\parindent}{0cm}

\begin{document}

\title{\textbf{KINC v0.1 Specification}\\
\vspace{1mm}
\includegraphics[width=8cm,height=4cm]{KINClogo.png}
\\Knowledge Independent Network Construction}
\author{Joshua Burns\thanks{Dept. of Horticulture, Washington State 
University}, Stephen Ficklin\footnotemark[1]}
\maketitle

\newpage
\tableofcontents

\newpage
\section{Introduction}

KINC is designed for use in construction of biological networks, specifically, 
gene co-expression networks.

KINC performs three major steps:

\begin{list}{}{}
\item[1)] Construction of a similarity matrix of pair-wise expression 
correlations.
\item[2)] Thresholding of the similarity matrix to form an adjaceny matrix.
\item[3)] Export of the adjaceny matrix to form a tab-delimited network file.
\end{list}

This document provides an overview for the data structures and file formats 
used by KINC.

The KINC program upon loading presents the user with a console. 
\autoref{intro:ccommands} shows the basic commands this console will support.

\begin{figure}[H]
\begin{mdframed}[style=btable]
\begin{tabularx}{\textwidth}{|l|X|}
\hline
\textbf{Command} & \textbf{Description} \\
\hline
history & Shows the history of a given data file. \\
\hline
load & Loads a new data file with the given type and name. \\
\hline
export & Exports a given data file to an ASCII export file. \\
\hline
query & Query a given data file for information. \\
\hline
merge & Merge two or more data files into one, if possible. \\
\hline
list & List all loaded data files. \\
\hline
gpu & GPU commands. \\
\hline
\end{tabularx}
\end{mdframed}
\caption{Basic Console Commands}
\label{intro:ccommands}
\end{figure}

Any other command will be treated as the name of an analytic object to load and 
execute. If no analytic is found with the given name nothing is done.

The GPU command has a list of subcommands for querying and setting up any GPGPU 
device to be used with analytic objects. \autoref{intro:gpucommands} shows all 
subcommands pertaining to GPGPU setup through OpenCL.

\begin{figure}[H]
\begin{mdframed}[style=btable]
\begin{tabularx}{\textwidth}{|l|X|}
\hline
\textbf{Command} & \textbf{Description} \\
\hline
list & List all availbale OpenCL compatible devices. \\
\hline
set & Set a specific OpenCL device to be used for accelerated computation. \\
\hline
clear & Clear any previously set OpenCL device. \\
\hline
\end{tabularx}
\end{mdframed}
\caption{GPU Console Subcommands}
\label{intro:gpucommands}
\end{figure}

\newpage
\section{Main Program}

The main program consists of a console management class which acts as the 
program's controller and two abstract data classes that create a common 
interface to the program any implementation object or analytic must inherit and 
follow.

\subsection{Console}

The \h{Console} class has a single instance within the main program and is 
designed to be given control through its \h{run()} 
function. This class creates a terminal console for the user, creates and 
manages all \h{Data} objects, and handles all \h{Analytic} execution.

This class also interfaces with the plugin object factory functions that 
generate new data or analytic interface object based off their unique names. 
These functions are implemented in a common source file which requires editing 
whenever a new plugin is added to the program.

\autoref{fig:console:func} shows the functions this class implements.

The constructor takes any command line arguments.

\begin{figure}[H]
\begin{mdframed}[style=functions]
\begin{lstlisting}[language=C++]
Console(int,char*[]);
void run();
bool list(Data*,std::string&);
bool unlist(std::string&);
Data* find(std::string&);
Data* new_data(std::string&);
Analytic* new_analytic(std::string&);
\end{lstlisting}
\end{mdframed}
\caption{Functions for Console Class}
\label{fig:console:func}
\end{figure}

The \h{run(...)} function takes control of the process and runs a console for 
the user or executes a script, depending on the command line arguments 
supplied. Once the user exits the console or the script has finished executing 
this function returns control to the caller.

The \h{list(...)} function adds a new \h{Data} object to the list of 
available \h{Data} objects within the console with the name string supplied. 
The name must be unique from all other loaded \h{Data} objects.

The \h{unlist(...)} function removes the \h{Data} object with the name 
string specified if it exists. If it exists and it was removed it returns 
\h{TRUE} else it returns \h{FALSE}.

The \h{find(...)} function finds a loaded \h{Data} object with the given name. 
If no object is found with that name then \h{NULL} is returned, else a 
pointer to the found object is returned.

The \h{new\_data(...)} function creates a new data object of the type 
specified in the supplied string. If the string is not a valid data type then 
\h{NULL} is returned, else a pointer to the new data object is returned.

The \h{new\_analytic(...)} creates a new analytic object of the type specified 
in the supplied string. If the string is not a valid data type then \h{NULL} is 
returned, else a pointer to the new analytic object is returned.

The \h{new\_data(...)} and \h{new\_analytic(...)} functions are designed to be 
plugin object factories which return data and analytic objects with the type 
given to them as a string. These two functions along with the list of all 
available plugins containing their unique string names and unique number 
identifiers are all contained in the same source and header file. These two 
files will represent where new static plugins that implement either a data or 
analytic interface can be added.

\subsection{Data}

The abstract \h{Data} class creates a common data object interface to the 
console program and provides file input/output for any data class implementing 
it. This class is responsible for implementing the basic file input and output 
operations along with specifying a common interface with the console program.

Additional functions can and should be added to any implementation of this 
interface class that deals with manipulating the specific type of data that is 
being implemented.

\autoref{fig:data:func} shows the public functions this class defines. 
Exluding the constructor, all of these functions are pure virtual functions 
that any class inheriting this class are required to implement. The constructor 
takes a single argument which is the binary file location where the data for an 
object is stored.

\begin{figure}[H]
\begin{mdframed}[style=functions]
\begin{lstlisting}[language=C++]
Data(const std::string&);
void __history();
virtual uint32_t type() = 0;
virtual bool __load(std::vector<std::string>&) = 0;
virtual bool __export(std::vector<std::string>&) = 0;
virtual bool __query(std::vector<std::string>&) = 0;
virtual bool __merge(std::vector<std::string>&) = 0;
virtual bool flush() = 0;
\end{lstlisting}
\end{mdframed}
\caption{Functions for Data Class}
\label{fig:data:func}
\end{figure}

The \h{\_\_history()} function prints the entire history of its object to 
standard output.

The virtual \h{type()} function must return the unique identier of the type of 
data this object represents. The list of unique identifiers are stored in a 
common header file.

The \h{\_\_load(...)}, \h{\_\_export(...)}, \h{\_\_query(...)}, and 
\h{\_\_merge(...)} functions are all supplied with the same argument. This 
argument is a list of additional arguments given to the specific command issued 
by the user on the console.

The virtual \h{\_\_load(...)} function is called when a load console command 
is issued on a new data object. The data object instance can assume that it is 
empty and was just created by the console. This function is responsible for 
loading data from a human readable ASCII file and encoding it into its binary 
format in a new file of the same name.

The virtual \h{\_\_export(...)} function is called when a export console command 
is issued on a given data object. This function is responsible for exporting 
its internal data stored in binary format and decoding it into a human readable 
ASCII file.

The virtual \h{\_\_query(...)} function is called when a query console command 
is issued on a given data object. This function is responsible for giving 
information about the data this object holds to the user based off arguments 
given.

The virtual \h{\_\_merge(...)} function is called when a merge console command 
is issued on a given data object. This function is responsible for merging two 
or more seperate data objects of the same type into a new single data object. 
The data object this command is called on is the first data object in the list 
of objects that will be merged. The rest of the objects are supplied as 
additional arguments. This function does not have to merge the given data 
objects if the data is of a type that cannot be merged without calling an 
analytic object.

The \h{\_\_load(...)}, \h{\_\_export(...)}, \h{\_\_merge(...)}, and \h{flush()} 
functions will not return control to the caller of the function until all write 
operations to the binary file the data object represents have completed.

\autoref{fig:data:func:prot} shows the protected functions this class 
implements that a class inheriting this abstract class can use for file input 
and output of its binary data. These functions are designed to hide the raw 
file from an implemenation class so it is impossible to overwrite the header 
part of the file that this abstract class is responsible for maintaining.

\begin{figure}[H]
\begin{mdframed}[style=functions]
\begin{lstlisting}[language=C++]
uint64_t fsize();
bool fgrow(uint64_t);
void fseek(uint64_t);
template<class T> bool fread(T*,uint64_t);
template<class T> bool fwrite(T*,uint64_t);
\end{lstlisting}
\end{mdframed}
\caption{Protected Functions for Data Class}
\label{fig:data:func:prot}
\end{figure}

The \h{fsize()} function returns the total size of the binary file in bytes. 
This does not include the header information that is hidden from any 
implementation class.

The \h{fgrow()} function will increase the size of the binary file by the 
number of bytes given. If this was successful it returns \h{TRUE} else if it 
could not grow the size of the file it returns \h{FALSE}.

The \h{fseek()} function will move the file position's indent to the number 
given in bytes. This number must be within range of the total size of the 
binary file.

The \h{fread(...)} function reads the number of elements given and writes them 
to the pointer of that element type given, starting at the current file 
position and incrementing by the number of bytes read. The new file position 
cannot exceed the total size of the file.

The \h{fwrite(...)} function overwrites the number of elements given from the 
given pointer to the binary file, starting at the current file position and 
incrementing by the number of bytes overwritten. The new file position cannot 
exceed the total size of the file.

This abstract class is responsible for reading in the header information of any 
data file since it is generic to any specific data type. This part of the 
binary file is hidden from any implementation class.

\autoref{fig:kincdata:header} shows the binary format for the beginning 
of a data file.

\begin{figure}[H]
\begin{mdframed}[style=btable]
\begin{tabularx}{\textwidth}{|l|l|X|l|}
\hline
\textbf{Name} & \textbf{Value} & \textbf{Description} & \textbf{Type} \\
\hline
header & KINC & Special header tag specifying this is a KINC binary data file. 
& char[4] \\
\hline
type & & Number that defines Data type for a file. & uint32\_t \\
\hline
historySize & & Total size of all history items in bytes. & uint32\_t \\
\hline
history & & Array of history items. & byte[historySize] \\
\hline
\end{tabularx}
\end{mdframed}
\caption{Binary File Format of Header}
\label{fig:kincdata:header}
\end{figure}

The \h{header} field is a special tag that specifies this is a KINC data file. 
The \h{type} field represents the specific type of data this file represents. 
The \h{historySize} field represents the total size of all history data in 
bytes.

\autoref{fig:kincdata:history} shows the format for a single history item 
within the history buffer. History items are nested inside one another, the 
highest history element being the history of the current file, and all 
subhistories being the history for all input files.

\begin{figure}[H]
\begin{mdframed}[style=btable]
\begin{tabularx}{\textwidth}{|l|X|l|}
\hline
\textbf{Name} & \textbf{Description} & \textbf{Type} \\
\hline
elemSize & The total size of this history item and all subhistory items it 
bytes. & uint64\_t \\
\hline
fileLen & Length of file name string in bytes. & uint16\_t \\
\hline
objectLen & Length of object name string in bytes. & uint16\_t \\
\hline
commandLen & Length of command string in bytes. & uint16\_t \\
\hline
date & Linux time-stamp of when file was last modified. & uint64\_t \\
\hline
subHistoryAmt & Number of input history items. & uint16\_t \\
\hline
subHistorySize & Size of input history data in bytes. & uint32\_t \\
\hline
file & File name string. & char[fileLen] \\
\hline
object & Name of object that built file in bytes. & char[objectLen] \\
\hline
command & Command line used in console to construct data file. & 
command[commandLen] \\
\hline
subHistory & Array of input history items. & byte[subHistorySize] \\
\hline
\end{tabularx}
\end{mdframed}
\caption{Binary File Format of Individual History Item}
\label{fig:kincdata:history}
\end{figure}

The \h{elemSize} field represents the total size of this history item, 
including all subhistory items nested within it.

The \h{file}, \h{object}, and \h{command} fields represent the name of the 
files, the object that created the data, and the specific KINC console command 
that invoked the creation of the file. For the \h{object} field, it begins with 
``D/'' or ``A/'' that states it was made by either a data object or an analytic 
object, respectively.

The \h{date} field represents the date when this data file was last modified.

The \h{subHistoryAmt} and \h{subHistorySize} represents the total number of 
history items in the list of subhistory items nested within this history item 
and the total size of all nested history items in bytes, respectively.

The \h{subHistory} field contains the list of all nested subhistory items 
contained within this history item.

\subsection{Analytic}

The abstract \h{Analytic} class creates a command analytic interface to the 
console program. This class is almost a pure virtual class and only supplies a 
pointer to an OpenCL context from its constructor, provided to it from the main 
console.

\autoref{fig:kincanalytic} shows the virtual functions an implemenation of this 
class must define along with its constructor. The constructor takes a pointer 
to an OpenCL context that can be used for accelerated computation of the data, 
but it is not required and can be given \h{NULL}. An implementation of this 
class is required to provide a means to compute its data with or without an 
OpenCL device.

\begin{figure}[H]
\begin{mdframed}[style=functions]
\begin{lstlisting}[language=C++]
Analytic(cl::Context*);
virtual uint32_t type() = 0;
virtual bool execute(std::vector<std::string>&) = 0;
\end{lstlisting}
\end{mdframed}
\caption{Functions for Analytic Class}
\label{fig:kincanalytic}
\end{figure}

The virtual \h{type()} function must return the unique identier of the type of 
analytic this object represents. The list of unique identifiers are stored in a 
common header file.

The virtual \h{execute(...)} function is called for this analytic to execute 
the task it is designed to implement. The argument passed to it is a list of 
additional arguments given to the specific command issued by the user on the 
console. This command should be blocking until all file input/output has 
completed.

Typically this class takes input from one or more data files and creates a new 
data file as a result. This class is responsible for communicating to these 
individual data objects and creating a new data object if one is to be created. 
This class must know all pertanent functions involved in all data types it is 
responsible for reading and writing to.


















\newpage
\section{Data Classes}

\subsection{Expression}

The Expression class is responsible for manging gene expression-level data.  

\subsubsection{Properties}

\stephen{Do we need any properties?}

\josh{I don't think we need properties. I am also unsure how to implement them 
in C++. My thought is all interactions between the classes will be defined in 
the Abstract Classes section using virtual functions?}

\subsubsection{Constructor}

{\bfseries Data(int argc, char *argv[])}

\stephen{We need to design how the functions of the class will receive 
arguments.  will we have a constructor that receives, parses and responds to 
errors for all functions?  Or should each function be responsible for checking 
it's own arguments. I know we can't do that in the abstract class, but 
we need to accomdate the behavior we settle on in our design so plugins
are consistent.}

\josh{I almost completely agree. These functions and interactions will all be 
defined in the abstract classes section if that is OK? It is standard C++ to 
define everything you are talking about in the abstract interface class with 
virtual functions. It is usually a good idea to have a default constructor only 
for implemenation classes, and have any additional configuation added into 
additional virtual functions that any implementation must handle.}

\subsubsection{Virtual Functions}

The following functions should be implemented by any plugin that creates
classes that inherits the Data class.

{\bfseries virtual void import() = 0}

This function reads a tab-delimited file.  Each line of this file
represents the gene expression levels of a single gene, transcript or probeset.
Each tab-separated value in a single line indicates the gene expression level 
for each sample. The expression level of a samples must be in the same 
order for every line.  The first line of the file may contain a tab-delimited
list of sample names, and a file may contain as many samples and genes as
desired.

\subsubsection{File Structure}

\autoref{fig:expression} shows the binary format of expression data and how it 
is stored on file. \h{geneAmt} and \h{sampleAmt} give the total number of genes 
and samples in the data, respectively. \h{geneNames} is the list of all gene 
names as a string who's length and partitioning is defined by \h{geneNameLen} 
and \h{geneNameSize}. \h{sampleNames} is the list of all sample names as a 
string who's length and partitioning is defined by \h{sampleNameLen} and 
\h{sampleNameSize}. Lastly, \h{samples} is 2 dimensional matrix of all samples 
for each gene, where the matrix is sorted by gene major order.

\begin{figure}[H]
\begin{mdframed}[style=btable]
\begin{tabularx}{\textwidth}{|l|X|l|}
\hline
\textbf{Name} & \textbf{Description} & \textbf{Type} \\
\hline
geneAmt & Total number of genes. & uint32\_t \\
\hline
sampleAmt & Number of samples per gene. & uint32\_t \\
\hline
geneNameLen & Length of each string identifying genes. & uint16\_t \\
\hline
geneNameSize & Total size of gene name list in bytes. & uint64\_t \\
\hline
sampleNameLen & Length of each string identifying samples. & uint16\_t \\
\hline
sampleNameSize & Total size of sample name list in bytes. & uint64\_t \\
\hline
geneNames & List of gene string identifiers. & char[geneNameSize] \\
\hline
sampleNames & List of sample string identifiers. & char[sampleNameSize] \\
\hline
sampleTotal & Total number of samples for all genes. & uint64\_t \\
\hline
samples & List of all samples per gene. & float[sampleTotal] \\
\hline
\end{tabularx}
\end{mdframed}
\caption{Binary File Format of Expression Data}
\label{fig:expression}
\end{figure}

\subsection{Correlation}

This is responsible for storing correlation data between genes.

The following describes the format of the KINC correlation file. All 
multi-byte numbers are little-endian, regardless of the machine endianness.

\stephen{I like this type of table for describing the file format. I borrowed 
it from the BAM file specification}

\josh{So do I! I was actually going to convert these defintions to a tabular 
format after your first review. :)}

\autoref{fig:correlation} shows the binary format of correlation data and how 
it is stored on file. \h{geneAmt}, \h{sampleAmt}, and \h{corrAmt} give the 
number of genes, number of samples per gene, and number of correlations per 
gene, respectively. \h{geneNames} is the list of all gene names that are 
correlated who's length and partitioning is defined by \h{geneNameLen} and 
\h{geneNameSize}. \h{sampleNames} is the list of all sample names used for 
correlation between genes who's length and partitioning is defiend by 
\h{sampleNameLen} and \h{sampleNameSize}. \h{corrTypes} is the list of all 
correlation types listed for all gene pairs who's length and partitioning is 
defined by \h{corrTypeLen} and \h{corrTypeSize}. Lastly, \h{correlations} is a 
special diagonal matrix where all correlations for gene pairs are stored using 
gene major order.

\begin{figure}[H]
\begin{mdframed}[style=btable]
\begin{tabularx}{\textwidth}{|l|X|l|}
\hline
\textbf{Name} & \textbf{Description} & \textbf{Type} \\
\hline
geneAmt & Total number of genes. & uint32\_t \\
\hline
sampleAmt & Number of samples per gene. & uint32\_t \\
\hline
corrAmt & Number of correlations per gene relationship. & uint8\_t \\
\hline
geneNameLen & Length of each string identifying genes. & uint16\_t \\
\hline
geneNameSize & Total size of gene name list in bytes. & uint64\_t \\
\hline
sampleNameLen & Length of each string identifying samples. & uint16\_t \\
\hline
sampleNameSize & Total size of sample name list in bytes. & uint64\_t \\
\hline
corrTypeLen & Length of each string identifying correlation type. & uint16\_t \\
\hline
corrTypeSize & Total size of correlation type list in bytes. & uint16\_t \\
\hline
geneNames & List of gene string identifiers. & char[geneNameSize] \\
\hline
sampleNames & List of sample string identifiers. & char[sampleNameSize] \\
\hline
corrTypes & List of correlation type strings. & char[corrTypeSize] \\
\hline
corrTotal & Total number of correlations for all gene relationships. & 
uint64\_t \\
\hline
correlations & Diagonal matrix list of all gene correlations for all 
relationships. & float[corrTotal] \\
\hline
\end{tabularx}
\end{mdframed}
\caption{Binary File Format of Correlation Data}
\label{fig:correlation}
\end{figure}

\subsection{Network}

This is responsible for storing network data between genes.

\autoref{fig:network} shows the binary format of network data and how it is 
stored on file. \h{geneAmt} give the number of genes in the network. 
\h{geneNames} is the list of all gene names that are correlated who's length 
and 
partitioning is defined by \h{geneNameLen} and \h{geneNameSize}. Lastly, 
\h{network} is a special diagonal matrix where all network edges for gene pairs 
are stored using gene major order.

\begin{figure}[H]
\begin{mdframed}[style=btable]
\begin{tabularx}{\textwidth}{|l|X|l|}
\hline
\textbf{Name} & \textbf{Description} & \textbf{Type} \\
\hline
geneAmt & Total number of genes. & uint32\_t \\
\hline
geneNameLen & Length of each string identifying genes. & uint16\_t \\
\hline
geneNameSize & Total size of gene name list in bytes. & uint64\_t \\
\hline
geneNames & List of gene string identifiers. & char[geneNameSize] \\
\hline
netTotal & Total number of edges, true or false, in network data. & uint64\_t \\
\hline
network & Diagonal matrix list of all possible edges in gene network. & 
bool[netTotal] \\
\hline
\end{tabularx}
\end{mdframed}
\caption{Binary File Format of Network Data}
\label{fig:network}
\end{figure}

\subsection{Annotation}

This is responsible for storing additional information for genes.

\autoref{fig:annotation} shows the binary format of annotation data and how it 
is stored on file. \h{geneAmt} and \h{annotAmt} give the number of genes and 
the number of annotations, respectively. \h{geneNames} is the list of all gene 
names who's length and partitioning is defined by \h{geneNameLen} and 
\h{geneNameSize}.\h{annotNames} is the list of all annotation names who's 
length and partitioning is defined by \h{annotNameLen} and \h{annotNameSize}.
\h{annotValSize} is a list of all string lengths for each annotation value per 
each gene. Lastly, \h{annotations} is a 2 dimensional matrix that lists all 
annotations for all genes using gene major order.

\begin{figure}[H]
\begin{mdframed}[style=btable]
\begin{tabularx}{\textwidth}{|l|X|l|}
\hline
\textbf{Name} & \textbf{Description} & \textbf{Type} \\
\hline
geneAmt & Total number of genes. & uint32\_t \\
\hline
annotAmt & Total number of annotations per gene. & uint32\_t \\
\hline
geneNameLen & Length of each string identifying genes. & uint16\_t \\
\hline
geneNameSize & Total size of gene name list in bytes. & uint64\_t \\
\hline
annotNameLen & Length of each string identifying the name of a annotation. & 
uint16\_t \\
\hline
annotNameSize & Total size of annotation name list in bytes. & uint64\_t \\
\hline
annotNames & List of annotation string identifiers. & char[annotNameSize] \\
\hline
geneNames & List of gene string identifiers. & char[geneNameSize] \\
\hline
annotValLens & List of numbers that identify the length of each value string 
for each annotation. & uint16\_t[annotAmt] \\
\hline
annotValSize & Total size of all annotation values. & uint64\_t \\
\hline
annotations & List of all annotations per gene. & char[annotValSize] \\
\hline
\end{tabularx}
\end{mdframed}
\caption{Binary File Format of Annotation Data}
\label{fig:annotation}
\end{figure}

\newpage
\section{Analytic Classes}

\subsection{Pearson}

This takes an Expression BioData object and produces a Correlation BioData 
object. It uses the Pearson correlation statistical method for giving 
correlation values.

\subsection{Spearman}

This takes an Expression BioData object and produces a Correlation BioData 
object. It uses the Spearman correlation statistical method for giving 
correlation values.

\end{document}
